### PR TITLE
refactor: create `ensure_igraph()`

### LINKED ISSUE #702: add a function assert_igraph()?
This is a non-critical refactoring suggestion: adding a function that asserts whether something is an igraph object, and errors if not. 

To replace lines such as

```r
  if (!is_igraph(graph)) {
    stop("Not a graph object")
  }
```

and the ones that also first checks whether graph is NULL.

### PR DESCRIPTION
**Pull request by maelle**
Fix #702

### CODE REVIEW

**Comment by maelle** [comment]
I'm realizing I don't know where aaa-auto.R comes from: should I fix the conflicts in that script or somewhere more upstream?

**Comment by maelle** [comment]
Another question: should `ensure_igraph()` actually error for NA and NULL?

**Comment by ntamas** [comment]
> Another question: should ensure_igraph() actually error for NA and NULL?

I think it should. If we need to support NULL/NA _and_ real graphs for an input argument, then we should have an  `ensure_optional_igraph()` function or an `ensure_igraph(allow_na=...)` variant at least.

**Comment by krlmlr** [comment]
`NA` should be an error IMO, `NULL` may be fine indeed if optional.

`aaa-auto.R` is generated by Stimulus, we should add a "do not edit by hand" header. It's fine if you keep that unchanged, we can adapt the code generation later.

**Comment by ntamas** [comment]
Is there a fine distinction between `NULL` and `NA` that I am not aware of? (Like between `undefined` and `null` in JS?)

**Comment by krlmlr** [comment]
`NA` is a logical vector of length one, `NULL` is a sentinel object. 

See https://design.tidyverse.org/def-short.html#arg-short-null and surrounding.

**Comment by maelle** [comment]
where should I make changes for them to be in R/aaa-auto.R?

**Comment by ntamas** [comment]
`tools/stimulus/aaa-auto.R.in` is the template that the code generator extends in order to produce `R/aaa-auto.R`, would that work?

**Comment by maelle** [comment]
`tools/stimulus/aaa-auto.R.in` is only two lines though :thinking:

**Comment by maelle** [comment]
I tried searching for occurrences of `gorder_impl` but I think I might be looking at the wrong repo, does code in another repo produce code for the rigraph repo?

**Comment by maelle** [comment]
oh I see types-RR.yaml

**Comment by maelle** [comment]
at the moment, the only places where I saw NULL is ok according to the argument check, don't seem they'd actually work with NULL: https://github.com/search?q=repo%3Aigraph%2Frigraph+is.null%28graph%29+language%3AR&type=code&l=R

**Comment by krlmlr** [comment]
Please change `aaa-auto.R` as you see fit, we'll adapt Stimulus later.

@ntamas: Are you fine with importing cli, which is itself dependency-free? It provides nicer error messages.

**Comment by krlmlr** [comment]
Did my best to resolve conflicts, sorry if I messed up

**Comment by ntamas** [comment]
> Are you fine with importing cli, which is itself dependency-free? It provides nicer error messages.

Sure, no problem.

**Comment by maelle** [comment]
How about https://github.com/igraph/rigraph/blob/7a70f0e7aa9733b3b462db9cceaaf900ef00e53c/R/scg.R#L709? Can one only supply a matrix?

File: R/attributes.R [hunk 1]
```diff
@@ -574,6 +556,9 @@ edge.attributes <- function(graph, index = E(graph)) {
 #' g <- make_ring(10)
 #' graph_attr_names(g)
 graph_attr_names <- function(graph) {
+  ensure_igraph(graph)
+
+  res <- .Call(C_R_igraph_mybracket2_names, graph, igraph_t_idx_attr, igraph_attr_idx_graph)
```
Comment (maelle) [inline R/attributes.R:561]: here too @krlmlr

File: R/attributes.R [hunk 2]
```diff
@@ -112,15 +110,15 @@ graph_attr <- function(graph, name) {
...
-  }
+  ensure_igraph(graph)
 
   .Call(R_igraph_mybracket3_set, graph, igraph_t_idx_attr, igraph_attr_idx_graph, name, value)
 }
 
 #' @export
 graph.attributes <- function(graph) {
+  ensure_igraph(graph)
+  .Call(C_R_igraph_mybracket2_copy, graph, igraph_t_idx_attr, igraph_attr_idx_graph)
```
[the remark this answers is not in the document]
Comment (maelle) [reply R/attributes.R]: I made a commit removing duplicate code

**Comment by ntamas** [comment]
> How about
> 
> https://github.com/igraph/rigraph/blob/7a70f0e7aa9733b3b462db9cceaaf900ef00e53c/R/scg.R#L709
> ? Can one only supply a matrix?

Nice catch. Indeed the underlying [`igraph_scg_adjacency` function](https://igraph.org/c/html/0.9.4/igraph-SCG.html#igraph_scg_adjacency) can work from a graph, a matrix or a sparse matrix, but it only needs exactly one of these.

Note that the spectral coarse graining functions (i.e. everything that's in [this chapter](https://igraph.org/c/html/0.9.4/igraph-SCG.html) were removed in igraph 0.10 due to concerns about maintainability (in other words, no one really knew from the current dev team how this piece of code works). They now live in a [separate project](https://github.com/igraph/igraph-scg). If we want to keep these after upgrading the C core to igraph 0.10, we need to link to both projects.

**Comment by ntamas** [comment]
`cli` import is okay for me.

**Comment by maelle** [comment]
> How easy is it to split these changes between autogenerated and manual parts?

You tell me :wink: It's still not 100% clear to me how autogeneration happens but I did modify `tools/stimulus/types-RR.yaml`, hopefully correctly?

**Comment by krlmlr** [comment]
My rhetoric question is a disguised form of saying "this PR is too big"... Let's see if the tests pass and merge, it looks good otherwise.

**Comment by ntamas** [comment]
LGTM.

**Comment by krlmlr** [comment]
Thanks!

**Pull request merged**